File: web/features.js
export function analyzeRMSAndPauses(stream, onStats){
  const ctx = new (window.AudioContext||window.webkitAudioContext)({ sampleRate: 16000 });
  const src = ctx.createMediaStreamSource(stream);
  const proc = ctx.createScriptProcessor(1024, 1, 1);
  let samples = 0; let sumsq = 0;
  let pauses=0, maxPause=0, curPause=0;
  const sr = ctx.sampleRate; const frame= Math.round(0.02*sr); const hop=Math.round(0.01*sr);
  let buf = new Float32Array(0);
  src.connect(proc); proc.connect(ctx.destination);
  const start = performance.now();
  proc.onaudioprocess = e => {
    const x = e.inputBuffer.getChannelData(0);
    // accumulate stats
    for(let i=0;i<x.length;i++){ const v=x[i]; sumsq+=v*v; samples++; }
    // slide for pauses
    const merged = new Float32Array(buf.length + x.length);
    merged.set(buf,0); merged.set(x,buf.length); buf=merged;
    while(buf.length>=frame){
      const seg=buf.subarray(0,frame);
      let s=0; for(let i=0;i<frame;i++) s+=seg[i]*seg[i];
      const energy = Math.sqrt(s/frame);
      if(energy < 0.01){ curPause++; } else { if(curPause*hop/sr>=0.5) pauses++; if(curPause>maxPause) maxPause=curPause; curPause=0; }
      buf = buf.subarray(hop);
    }
    const dur = (performance.now()-start)/1000;
    const rms = 20*Math.log10(Math.max(1e-12, Math.sqrt(sumsq/Math.max(1,samples))));
    onStats({durationSec: dur, rmsDb: rms, pauses, maxPauseSec: maxPause*hop/sr});
  };
  return { stop(){ proc.disconnect(); src.disconnect(); ctx.close(); } };
}

export function charsPerMinute(refLen, durationSec){
  if(!durationSec) return 0; return 60*refLen/durationSec;
}

export function scoreFromMetrics(cer,cpm,pauses){
  const acc=Math.max(0,Math.min(100,(1-cer)*100));
  const low=200, high=350; let spd;
  if(cpm>=low && cpm<=high) spd=100; else if(cpm<low) spd=Math.max(0,100-(low-cpm)*0.5); else spd=Math.max(0,100-(cpm-high)*0.5);
  const pausePenalty=Math.min(10,Math.max(0,pauses));
  const pauseScore=Math.max(0,100-pausePenalty);
  const overall=0.7*acc+0.2*spd+0.1*pauseScore;
  return {accuracyScore:round1(acc),speedScore:round1(spd),pauseScore:round1(pauseScore),overallScore:round1(overall)};
}

function round1(x){return Math.round(x*10)/10}

File: ondoku_judge/features.py
from dataclasses import dataclass
from typing import Dict, Any
import numpy as np
import soundfile as sf


@dataclass
class AudioStats:
    duration_sec: float
    rms_db: float
    pauses: int
    max_pause_sec: float


def _rms_db(x: np.ndarray) -> float:
    eps = 1e-12
    rms = float(np.sqrt(np.mean(np.square(x) + eps)))
    return 20.0 * np.log10(max(rms, eps))


def _count_pauses(x: np.ndarray, sr: int, th: float = 0.01, min_pause: float = 0.5) -> (int, float):
    """しきい値th未満の無音区間を数える。min_pause秒以上をポーズ扱い。"""
    frame = int(0.02 * sr)  # 20ms
    hop = int(0.01 * sr)    # 10ms
    if len(x) < frame:
        return 0, 0.0
    energies = []
    for i in range(0, len(x) - frame + 1, hop):
        seg = x[i : i + frame]
        energies.append(np.sqrt(np.mean(seg * seg)))
    energies = np.asarray(energies)
    mask = energies < th
    # 連続trueの長さを秒に換算
    if not mask.any():
        return 0, 0.0
    pauses = 0
    maxlen = 0
    cur = 0
    for v in mask:
        if v:
            cur += 1
            maxlen = max(maxlen, cur)
        else:
            if cur > 0:
                dur = cur * hop / sr
                if dur >= min_pause:
                    pauses += 1
            cur = 0
    if cur > 0:
        dur = cur * hop / sr
        if dur >= min_pause:
            pauses += 1
        maxlen = max(maxlen, cur)
    max_pause = maxlen * hop / sr
    return pauses, max_pause


def analyze_audio(wav_path: str) -> AudioStats:
    x, sr = sf.read(wav_path, dtype="float32")
    if x.ndim == 2:
        x = x.mean(axis=1)
    duration = len(x) / sr if sr > 0 else 0.0
    rms = _rms_db(x)
    pauses, max_pause = _count_pauses(x, sr)
    return AudioStats(duration_sec=duration, rms_db=rms, pauses=pauses, max_pause_sec=max_pause)


def compute_speed_chars_per_minute(ref_len_chars: int, duration_sec: float) -> float:
    if duration_sec <= 0:
        return 0.0
    return 60.0 * ref_len_chars / duration_sec


def score_from_metrics(cer: float, cpm: float, pauses: int) -> Dict[str, Any]:
    # 正確さスコア
    acc = max(0.0, min(100.0, (1.0 - cer) * 100.0))
    # 速度スコア（200–350字/分を満点帯とし、外側は線形減点）
    low, high = 200.0, 350.0
    if cpm >= low and cpm <= high:
        spd = 100.0
    elif cpm < low:
        spd = max(0.0, 100.0 - (low - cpm) * 0.5)
    else:  # cpm > high
        spd = max(0.0, 100.0 - (cpm - high) * 0.5)
    # ポーズ（1回につき-1点、上限-10）
    pause_penalty = min(10.0, float(max(0, pauses)))
    pause_score = max(0.0, 100.0 - pause_penalty)
    # 総合（重み付け）
    overall = 0.7 * acc + 0.2 * spd + 0.1 * pause_score
    return {
        "accuracy_score": round(acc, 1),
        "speed_score": round(spd, 1),
        "pause_score": round(pause_score, 1),
        "overall_score": round(overall, 1),
    }

